add.go

```diff
@@ -776,6 +856,123 @@ func (b *Builder) Add(destination string, extract bool, options AddAndCopyOption
+		if err != nil {
+			return fmt.Errorf("error getting container directory for temporary file: %w", err)
+		}
+
+		tempLayerDir := filepath.Join(containerDir, "link-layer-temp")
+		if err := os.MkdirAll(tempLayerDir, 0o700); err != nil {
+			return fmt.Errorf("error creating temporary directory for --link layer: %w", err)
+		}
+
+		tempLayerFile, err := os.CreateTemp(tempLayerDir, "layer-*.tar")
+		if err != nil {
+			return fmt.Errorf("error creating temporary file for --link layer: %w", err)
+		}
+
+		tempFileName := tempLayerFile.Name()
+
+		isDirectory := false
+		if len(sources) > 0 && !sourceIsRemote(sources[0]) && !sourceIsGit(sources[0]) {
+			if st, err := os.Stat(sources[0]); err == nil && st.IsDir() {
+				isDirectory = true
+			}
+		}
+
+		if isDirectory {
+			b.ContentDigester.Start("dir")
+		} else {
+			b.ContentDigester.Start("file")
+		}
+
+		hasher := b.ContentDigester.Hash()
+		if hasher == nil {
+			tempLayerFile.Close()
+			os.Remove(tempFileName)
+			return fmt.Errorf("hasher returned nil from ContentDigester.Hash()")
+		}
+		defer hasher.Close()
+
+		multiWriter := io.MultiWriter(hasher, tempLayerFile)
+
+		tarOpts := &archive.TarOptions{
+			Compression: archive.Uncompressed,
+		}
+
+		tarstream, err := chrootarchive.Tar(stagingDir, tarOpts, stagingDir)
+		if err != nil {
+			tempLayerFile.Close()
+			os.Remove(tempFileName)
+			return fmt.Errorf("error tarring staging directory %q: %w", stagingDir, err)
+		}
+		defer tarstream.Close()
+
+		if _, err := io.Copy(multiWriter, tarstream); err != nil {
+			tempLayerFile.Close()
+			os.Remove(tempFileName)
+			return fmt.Errorf("hashing and writing final layer tar: %w", err)
+		}
+
+		tempLayerFile.Close()
+
+		contentType, digest := b.ContentDigester.Digest()
+
+		linkLayerDir := filepath.Join(containerDir, "link-layers")
+		if err := os.MkdirAll(linkLayerDir, 0o700); err != nil {
+			os.Remove(tempFileName)
+			return fmt.Errorf("creating link-layers directory: %w", err)
+		}
+
+		layerFile := filepath.Join(linkLayerDir, fmt.Sprintf("%s-%s.tar", digest.Algorithm().String(), digest.Hex()))
+
+		if err := os.Rename(tempFileName, layerFile); err != nil {
+			os.Remove(tempFileName)
+			return fmt.Errorf("moving tempfile to final location %q: %w", layerFile, err)
+		}
+
+		var created time.Time
+		if options.Timestamp != nil {
+			created = options.Timestamp.UTC()
+		} else {
+			created = time.Now().UTC()
+		}
+
+		Command := "ADD"
+		if !extract {
+			Command = "COPY"
+		}
+
+		summary := contentType
+		if digest != "" {
+			if summary != "" {
+				summary = summary + ":"
+			}
+			summary = summary + digest.Encoded()
+			logrus.Debugf("added content from --link %s", summary)
+		}
+
+		createdBy := fmt.Sprintf("/bin/sh -c #(nop) %s --link %s in %s", Command, summary, destination)
+		history := v1.History{
+			Created:   &created,
+			CreatedBy: createdBy,
+			Comment:   b.HistoryComment(),
+		}
+
+		linkedLayer := LinkedLayer{
+			History:  history,
+			BlobPath: layerFile,
```

Comment:
This could point directly to `stagingDir`, skipping a lot of the work being done in between.